Merge v2018.01.2

dpino opened the pull request:
Merges v2018.01.2 into lwaftr branch.

alexandergall committed:
Add a Snabb program to construct a complete L2VPN endpoint

program/l2vpn creates a set of apps that implements an arbitrary
number of VPLS instances sharing a single uplink to the Internet from
a single configuration statement.  A VPLS instance is uniquely
identified by an IPv6 address that serves as the destination for all
pseudowires associated with it.  A VPLS can have any number of
customer ports ("attachement circuits").  A multi-point VPLS uses an
instance of a flooding or learning bridge connected to a full mesh of
pseudowires.

alexandergall committed:
Fix identification of proper p2p VPN

The logic that propagates the name of the AC interface to the
pseudowire module for inclusion in the service-specific MIB was flawed
since it did not check that the number of pseudowires.

alexandergall committed:
Add documentation

This is a first attempt at providing a user-friendly documentation.
It should be complete apart from some details (obviously missing
partes are marked with TODO)

alexandergall committed:
Fix bridge post_config() call

alexandergall committed:
Untabify apps/intel/intel1g.lua

alexandergall committed:
Add configurable MTU to intel1g.lua

Packets which exceed the MTU are dropped on ingress and egress.
Ingress drops are performed by the NIC in hardware (note that this
commit changes the "Store Bad Package" setting from enabled to
disabled, which is the default, to achieve this behaviour).  Egress
drops are performed in software.

Ingress drops are counted by the RLEC statistics register but are
currently not exposed by the driver.  Egress drops are not counted at
all.

alexandergall committed:
Add SNMP support to intel1g.lua

alexandergall committed:
Fix regression introduced by 11a19ce1

alexandergall committed:
Amend the list of devices supported by the Intel i210 driver

Also identify model-dependent code by the name of the model rather
than the device ID in intel1g.lua.

alexandergall committed:
Use /tmp to store test data in pseudowire selftest

alexandergall committed:
Fix ifSpeed in intel10g and intel1g

alexandergall committed:
Use logger:can_log() in l2vpn tunnel modules

This avoids NYIs for rate-limited log messages.

alexandergall committed:
Improve performance of the VLAN multiplexer

The performance of the VlanMux app (apps.vlan.vlan) is improved by

  - avoiding use of pairs()
  - avoiding string composition
  - serialisation of loops to avoid branches

The forwarding strategy during de-multiplexing has been changed to
forward all packets from the trunk links and accepting drops on vlan
links which are full (instead of only forwarding as many packets as
the fullest vlan link can accept).

OTOH, packets are retained on the incoming vlan links if the trunk
link is congested.  This strategy should probably be analyzed more
closely.

Support for different encapsulation schemes has been added: dot1q,
dot1ad and an explicit ethertype to cover non-standard encapsulations.

alexandergall committed:
Set ifAlias, ifDescr in intel drivers

Use explicit configuration of the ifAlias and ifDescr objects.  In
particular, ifAlias is not derived from the PCI address (except as
fallback) to allow for future naming schemes.

alexandergall committed:
Add oversized packets counter to ifInDiscards for intel10g

alexandergall committed:
Add app for static ND

apps.ipv6.nd_static fills in static src/dst MAC addresses in packets
transmitted on a L3 port.  Together with the nd_light and ns_responder
apps, it covers all cases of bidirectional dynamic ND, inbound
dynamic, outbound static and bidirectional static neighbor discovery.

The ns_responder app has been adapted to optionally fill in the
src/dst MAC addresses as well to allow for upper-layer protocols to be
agnostic of MAC addresses no matter which ND discipline is being used
on the interface.

alexandergall committed:
Add VLAN support to the l2vpn program

All uplinks and attachment circuits can now be configured to use
VLANs.  Uplinks are now a property of a VPLS instance instead of a
global property.  Interface specifications have been moved out of the
VPLS configuration and are structured much more like in a conventional
networking device.  Each interface now has distinct L2 and L3
configurations.  In case of a VLAN trunk, there can be any number of
subinterfaces representing distinct L3 interfaces.

alexandergall committed:
Remove MAC addresses from pseudowire app

The MAC addresses are now set by a suitable ND app in all cases.  The
pseudowire just needs to include the Ethernet header in the
encapsulation but the MAC addresses have no significance.

alexandergall committed:
Allow per-PW control-channel disable override

The cc can be disabled by setting heartbeat = 0.  This allows
disabling the cc for a PW when the default is enabled.

alexandergall committed:
Add missing image files

alexandergall committed:
Fix bug in l2vpn

The name of the interface app was set incorrectly in case of
an attachment circuit in non-trunking mode.

lukego committed:
Merge tag 'v2016.06' into vpn

alexandergall committed:
Remove SNMP code from Intel drivers

The new counter framework from v2016.08 has factored out the
interface-specific SNMP code to lib/ipv/shmem/iftable_mib.lua.  The
intel1g driver does not yet have the counter code in v2016.08,
probably due to an imminent major overhaul.

alexandergall committed:
Extract physical MAC address in intel10g

Make the physical MAC address available after initialization of the
driver.  This is used to populate the ifPhysAddress SNMP object and
supply the ND modules with the local address.

alexandergall committed:
Make ND apps reconfigurable

Add a reconfig() method to apps/ipv6/nd_{light,static} and
apps/ipv6/ns_responder.  This is primarily used to set the local MAC
address, which is only available until after the interface drivers
have been initialized.

alexandergall committed:
Fixes for new packet API

alexandergall committed:
Fixes/enhancements for iftable_mib.lua

  * add new iftypes
  * pass if{Descr,Name,Alias}
  * fix typos
  * use {r,t}xpackets as unicast packets if
    mcast/bcast stats are not available

alexandergall committed:
Streamline SNMP config in l2vpn

With the new counter framework in interface drivers and the
interface-specific SNMP code factored out to
lib/ipc/shmem/iftable_mib.lua it is possible to move the SNMP
configuration from the interface-level to the global config level and
add SNMP for sub-interfaces.

The README has been updated to reflect the new mode of configuration.

alexandergall committed:
Merge tag 'v2016.08' into l2vpn

alexandergall committed:
De-couple SNMP setup from ND reconfig

The re-configuration of the ND apps with the proper MAC addresses was
erroneously combined with the setting up of SNMP for interfaces,
causing ND to fail when SNMP was disabled.

Also fix a few glitches in the SNMP setup code wrt sub-interfaces.

alexandergall committed:
Augment stats counters for apps.tap.tap

Add counters for interface type, MAC address, MTU and operational
status to the apps.tap.tap app.  Also accept the MTU as a
configuration parameter to check whether the MTU configured on
the Snabb side matches the one configured on the TAP device to
avoid MTU mismatches.

alexandergall committed:
Add SIOCGIF{FLAGS,MTU,HWADDR} ioctls to ljsyscall

alexandergall committed:
Properly count the number of pseudowires

The number of pseudowires was not counted properly when determining
whether a VPN was p2p.  This caused the
cpwVcEnetPortIfIndex/pwEnetPortIfIndex objects to be set to an empty
string, which caused the pseudowire SNMP sub-agent to fail.

Also, the name of the interface was not set correctly.

alexandergall committed:
Disable interrupts in intel10g

alexandergall committed:
Make waiting for link-up optional in intel{1,10}g

This allows apps to complete startup even if a link doesn't come up
after initialisation.  The app needs to be able to cope with this
situation.

alexandergall committed:
nd_light: add option to reduce verbosity

The "quiet" configuration option is added to suppress "sending
neighbor solicitation" messages.

Commit:
ipsec test vs qemu works with vanfstd/snabblab-nixos:ipsec-linux-test (aad3824a2a703ff01bd44a5f065faee7d430cc2e)
command: nix-build -K jobsets/ipsec.nix -A tests --arg snabbSrc /path/to/snabb

Commit:
Fix minor things that crossed my path

alexandergall committed:
Fix program/l2vpn/pseudowire selftest

The pseudowire module now requires a nd_* module for setting the MAC
addresses on the uplink.

alexandergall committed:
Merge branch 'v2017.04' into l2vpn

Keep apps/intel/intel1g.lua for now until the l2vpn program has been
migrated to intel_mp

alexandergall committed:
Fix l2vpn pseudowire for new packet semantics

The "headroom" feature of a packet breaks the assumption that the
location of a packet is fixed.

alexandergall committed:
Fix tap after merge

   * Replace lib.timer by lib.throttle
   * Replace self.sock by self.fd in read()

alexandergall committed:
intel1g: replace lib.timer by lib.throttle

alexandergall committed:
nd_light: add "quiet" option to config template

The default for this option is false.

alexandergall committed:
inetl_app: add "wait_for_link" option to config template

The default for this option is false.

alexandergall committed:
l2vpn: switch to intel_mp drivers

alexandergall committed:
l2vpn: add extra_config to driver configuration

The configuration of a driver can have an optional element
extra_config, which must be a table and will be merged with the
regular config table.  This feature is intended to be used by
Lua-agnostic front-ends to the l2vpn program that create the
configuration from a higher layer of abstraction.

alexandergall committed:
l2vpn/README.md: Fix indention

alexandergall committed:
l2vpn: document new extra_config driver configuration item

Fix some indentions in inline code.

alexandergall committed:
l2vpn: use intel_mp in README throughout

alexandergall committed:
Remove intel1g driver

The intel1g driver was temporarily kept in the last merge with the
upstream master because it was required by the l2vpn program.  l2vpn
has been migrated to intel_mp and the intel1g driver can be retired
for good.

alexandergall committed:
Revert intel_app, inetl10g to upstream/master

l2vpn no longer supports the old driver.  The modifications on the
l2vpn branch are now obsolete and a merge with upstream would be
pointless.

alexandergall committed:
intel_mp: move shm initialization to constructor

Initialisation of the shm frame needs to be done in the constructor
method to avoid being overwritten by the frame created in core.app.

Add a copy of the mtu to the device-specific stats table, add a
counter to record the speed of the interface.

alexandergall committed:
iftable_mib: streamline processing

Move setting of interface speed to the periodic table update.

alexandergall committed:
Tap: add support for MTU, operational status

Also distinguish between ephemeral and persistent tap devices,
store the interface's MAC address in the stats shmem frame.

alexandergall committed:
l2vpn: access interface counters via read-only shm frames

Instead of accessing the private shm frames that hold the stats
counters of the devices, use copies via open_frame().

alexandergall committed:
apps.vlan: use config parser, update README

Also check the validity of the tag configuration option.

alexandergall committed:
nd_light: add optional option local_mac

The new option allows to statically configure the MAC address of the
next-hop IPv6 address for an environment where dynamic ND isn't
available.

alexandergall committed:
nd_light: fix bug in NS validation

The check for the solicited flag in a NS was not performed correctly.

alexandergall committed:
program.l2vpn.l2vpn: only use nd_light

With the inclusion of a static neighbor cache entry for the next-hop
address with 371e58, nd_light can perform all functionality previously
provided by nd_static and ns_responder.  To simplify the app, only
nd_light is used to implement the ND functionality.

The nd_static app is no longer needed and has been removed.  The
ns_responder app is no longer used either and is reset to the version
on master to simplify merging.

The l2vpn app no longer supports disabling replies to neighbor
solicitations when static next-hop resolution is enabled
("neighbor_nd" configuration option).

The "neighbor_mac" option has been renamed to "next_hop_mac".

alexandergall committed:
l2vpn: support dynamic reconfiguration

The l2vpn program supports a crude form of dynamic reconfiguration,
where the engine is run for a selectable interval after which the
configuration file is checked for changes.  If changes are detected,
the app network is reconfigured before the engine is restarted.

This kind of reconfiguraion is rather disruptive and it is probably
preferrable to restart the entire program when the configuration
changes until a smoother mechanism is developped.

alexandergall committed:
core.app: fix bug in reconfiguration

If an app network is re-configured, the 'keep' action copies the
existing input and output tables, but the code that processes the
links assumes that these tables are empty in the new app table.  This
causes the array-part of the tables to be duplicated.

alexandergall committed:
l2vpn: add port-mirroring option

Packets received or transmitted by an interface can be optionally
mirrored to either a Tap device or a pcap file for debugging purposes.

alexandergall committed:
Merge tag 'v2017.07' into l2vpn

alexandergall committed:
apps.vlan.vlan: remove assert in VlanMux link()

The new semantics of the link() method no longer allow this kind of
check, because it is called for each link action separately and there
is no way of knowing when all links have been configured.  The
asserttions for the existence of the trunk port is moved to push().

Also remove backpressure in the forwarding loops.

alexandergall committed:
apps.ipv6: document quiet option of nd_light

alexandergall committed:
l2vpn: re-write of app network construction

Yet another attempt to make this part of the code easier to read and
maintain.  It should also make it easier to move to a leader/follower
confgiruation using apps.config.

alexandergall committed:
l2vpn: fix bug in app links generation

The PW-facing links of the dispatcher apps need to be qualified by the
name of the VPLS because the PW names are only unique within a VPLS.

eugeneia committed:
lib.ipsec.esp: add support for tunnel mode

 - make lib.ipsec.esp AF independent in spirit
 - rename ESP_NH to PROTOCOL and make it public

eugeneia committed:
core.packet: resize should return the packet like the other functions

eugeneia committed:
lib.ipsec: update documentation w/r to tunnel mode and new interface

eugeneia committed:
program/snabbnfv/selftest.sh: move some functions to test_env for reuse

eugeneia committed:
Cleanup ipsec esp linux interoperability test.

eugeneia committed:
Merge branch 'master' into ipsec-linux-test

# Conflicts:
#	src/apps/tap/tap.lua

eugeneia committed:
apps/ipsec/selftest.sh: remove leftovers

eugeneia committed:
lib.ipsec.esp.decrypt_payload: properly handle packet underflow

Signal failure via return value instead of throwing an error.

eugeneia committed:
Merge tag 'v2017.08' into esp-tunnel

# Conflicts:
#	src/lib/ipsec/esp.lua

alexandergall committed:
Fix bug in p2p optimization

The name of the AC interface was not correctly referenced in the code
that removes the bridge module for p2p VPNs.

Also, the normalized name was not set for sub-interfaces.

eugeneia committed:
lib.ipsec.aes_128_gcm: use consttime_memequal to compare HMAC

- import consttime_memequal from NetBSD 7, a memory comparison
  predicate suitable for comparing inputs to secrets without leaking
  timing information
- use consttime_memequal in aes_128_gcm to verify authentication codes,
  plugging a timing leak

eugeneia committed:
lib.ipsec.aes_128_gcm: use fixed length ASM routine instead of consttime_memequal

This reverts commit b3072ab097ed3a38cba2b7c465eb88c00dd8f35c.

eugeneia committed:
apps/ipsec/selftest.sh: add environment variable for skipping the e2e test

wingo committed:
Enhance ptree readme with multi-process example

wingo committed:
Finish "snabb ptree" docs

wingo committed:
Merge pull request #1257 from Igalia/config-refactor

Flesh out multi-process reconfigurable network functions

wingo committed:
Fix "snabb lwaftr run" after lib.scheduling refactor

eugeneia committed:
lib.ipsec.esp.decrypt_payload: remove redundant bounds/spi checks

dpino committed:
Fix lwaftr selftest run

eugeneia committed:
core.link: free any packets left in link when freeing link

Due to how the engine works, the case that a link is non-empty between
breaths is rare (specifically, it should only occur when an app pushes
to a “slow” output interface). That’s probably why this went unnoticed
until now. I don’t even have a case that triggers this bug, it just came
to me.

eugeneia committed:
top: fix bug where counters were redundantly reopened

The formulation that was supposed to re-open counters on configuration
changes actually triggered re-opening on every polling.

alexandergall committed:
Add -jopt option to l2vpn program to set JIT options

Also increase default mcode limits (sizemcode=256, maxmcode=2048).

alexandergall committed:
Merge tag 'v2017.11' into l2vpn

eugeneia committed:
Merge PR #1261 (Merge l2vpn topic branch) into max-next

eugeneia committed:
src/README.md: improve docs of core packet ops that clobber their input value

The previously used term “ownership” was confusing because it is only
defined in terms of apps transfering packet ownership via links. This
new phrasing states that the affected operations “free” their packet
argument which should be intuitively understandable by most programmers,
especially given the explanation of packet.free.

eugeneia committed:
Merge branch 'esp-tunnel-docs' into esp-tunnel

eugeneia committed:
apps.ipsec: fixup incompatibilities introduces in esp-tunnel branch

eugeneia committed:
Merge tag 'v2017.11' into ipsec

eugeneia committed:
Merge branch 'esp-tunnel' into ipsec

* adds support for tunnel mode to lib.ipsec.esp
* clarifies core.packet documentation of operations that free their arguments

eugeneia committed:
Revert "src/README.md: improve docs of core packet ops that clobber their input value"

This reverts commit f0b2060c6fe18a7113fbd724388d5578412cb8d1.

eugeneia committed:
lib.ipsec.esp: paddle back on "freeing" vs "ownership" in documentation.

dpino committed:
Fix property-based test stall

lukego committed:
Merge branch master (v2017.11) into next

lukego committed:
.version: bump to v2018.01

wingo committed:
Merge remote-tracking branch 'origin/next' into wingo-next

lukego committed:
Merge #1256 branch 'snabbco/wingo-next' into next

eugeneia committed:
Merge PR #1260 (top: fix bug wrt counter reopening) into max-next

eugeneia committed:
Merge PR #1259 (core.link: free packets during link.free) into max-next

eugeneia committed:
Merge PR #1220 (IPsec ESP interoperability test with Linux) into max-next

lukego committed:
Merge #1265 branch 'snabbco/max-next' into next

eugeneia committed:
lib.lpm.lpm4: skip intensive test by default

This hides the more thorough LPM4 tests behind a flag / environment
variable SNABB_LPM4_TEST_INTENSIVE which can be set while hacking
on lib.lpm. For regular testing, this will significantly decrease the
total runtime of `make test'.

eugeneia committed:
Merge PR #1266 (ipsec: ESP tunnel mode and bug fixes) into max-next

eugeneia committed:
lib.numa: skip selftest if running on a non-NUMA machine

eugeneia committed:
Merge branch 'economic-selftests' into max-next

eugeneia committed:
Revert "lib.numa: skip selftest if running on a non-NUMA machine"

This reverts commit 2c71c239ce35baa943a8d2b25605250267878d91.

lukego committed:
Merge #1267 branch 'snabbco/max-next' into next

eugeneia committed:
Merge PR #1270 (v2018.01 release) into master

gdyr committed:
Naive i350 fixes

gdyr committed:
queue_registers, not registers

gdyr committed:
Fixed bits() on RO register

eugeneia committed:
apps.ipv4.reassemble: fix a copy paste error at end of selftest

eugeneia committed:
Merge tag 'v2018.01' into max-next

eugeneia committed:
Merge PR #1273 (lwaftr: fix test stall) into max-next

eugeneia committed:
Merge PR #1277 (i350 fixes) into max-next

eugeneia committed:
Merge PR #1280 (ipv4.reassemble: fix a copy paste error) into max-next

eugeneia committed:
apps.ipv6.reassemble: same for ipv6

eugeneia committed:
Merge PR #1280 (ipv4.reassemble: fix a copy paste error) into max-next

(Same for ipv6.reassemble)

eugeneia committed:
Merge PR #1282 (Fix lwaftr selftest run) into max-next

eugeneia committed:
Merge PR #1283 (Fix lwaftr loadtest) into max-next

gdyr committed:
Detect PCI -> port mapping

gdyr committed:
Merge branch 'i350-fixes' of https://github.com/gmichael225/snabb into i350-fixes

eugeneia committed:
Merge PR #1277 (Detect PCI -> port mapping) into max-next

eugeneia committed:
Merge branch 'master' into fixes

eugeneia committed:
doc/branches.md: take over fixes maintainership from lukego

eugeneia committed:
Merge branch 'max-next' (v2018.01.1 bugfix release) into fixes

This includes various fixes to regressions that caused test failures with the
CI, as well as a fix for intel_mp that makes it work with a specific flavor of
i350 NICs.

eugeneia committed:
Merge PR #1285 (v2018.01.1 bugfix release) into master

wingo committed:
Disable lwAFTR property-based tests unless SNABB_ENABLE_FLAKY_TESTS is set

Sadly, these tests are flaky and its failures are gating merges of
unrelated patches.  We'll work on fixing these tests on Igalia's lwaftr
branch.

eugeneia committed:
Merge PR #1294 (Disable lwAFTR property-based tests...) into fixes

eugeneia committed:
Merge PR #1298 (v2018.01.2 bugfix release) into master

dpino committed:
Merge tag 'v2018.01.2' into merge

dpino commented:
About performance, I see there are packet drops before reaching the 9.2 gbps packet load, but then it recovers.

```
Applying 9.000000 Gbps of load.             
  NIC 0:                                    
    TX 1653114 packets (1.653114 MPPS), 932356296 bytes (7.776248 Gbps)                 
    RX 1633190 packets (1.633190 MPPS), 986446760 bytes (8.205147 Gbps)                 
    Loss: 190 ingress drop + 19734 packets lost (1.205241%)                             
  NIC 1:                                    
    TX 1653114 packets (1.653114 MPPS), 998480856 bytes (8.305245 Gbps)                 
    RX 1633195 packets (1.633195 MPPS), 921121980 bytes (7.682549 Gbps)                 
    Loss: 184 ingress drop + 19735 packets lost (1.204938%)                             
Applying 9.200000 Gbps of load.             
  NIC 0:                                    
    TX 1653828 packets (1.653828 MPPS), 932758992 bytes (7.779607 Gbps)                 
    RX 1653828 packets (1.653828 MPPS), 998912112 bytes (8.308832 Gbps)                 
    Loss: 0 ingress drop + 0 packets lost (0.000000%)                                   
  NIC 1:                                    
    TX 1653828 packets (1.653828 MPPS), 998912112 bytes (8.308832 Gbps)                 
    RX 1653828 packets (1.653828 MPPS), 932758992 bytes (7.779607 Gbps)                 
    Loss: 0 ingress drop + 0 packets lost (0.000000%)                                   
Applying 9.400000 Gbps of load.             
  NIC 0:                                    
    TX 1652706 packets (1.652706 MPPS), 932126184 bytes (7.774329 Gbps)                 
    RX 1620991 packets (1.620991 MPPS), 979078564 bytes (8.143859 Gbps)                 
    Loss: 151 ingress drop + 31564 packets lost (1.918974%)                             
  NIC 1:                                    
    TX 1652706 packets (1.652706 MPPS), 998234424 bytes (8.303195 Gbps)                 
    RX 1620977 packets (1.620977 MPPS), 914231028 bytes (7.625076 Gbps)                 
    Loss: 142 ingress drop + 31587 packets lost (1.919821%)                             
Applying 9.600000 Gbps of load.             
  NIC 0:                                    
    TX 1650870 packets (1.650870 MPPS), 931090680 bytes (7.765692 Gbps)                 
    RX 1650870 packets (1.650870 MPPS), 997125480 bytes (8.293971 Gbps)                 
    Loss: 0 ingress drop + 0 packets lost (0.000000%)                                   
  NIC 1:                                    
    TX 1650870 packets (1.650870 MPPS), 997125480 bytes (8.293971 Gbps)                 
    RX 1650870 packets (1.650870 MPPS), 931090680 bytes (7.765692 Gbps)                 
    Loss: 0 ingress drop + 0 packets lost (0.000000%)                                   
Applying 9.800000 Gbps of load.             
  NIC 0:                                    
    TX 1630470 packets (1.630470 MPPS), 919585080 bytes (7.669731 Gbps)                 
    RX 1619155 packets (1.619155 MPPS), 977969620 bytes (8.134635 Gbps)                 
    Loss: 0 ingress drop + 11315 packets lost (0.693972%)                               
  NIC 1:                                    
    TX 1630470 packets (1.630470 MPPS), 984803880 bytes (8.191481 Gbps)                 
    RX 1619151 packets (1.619151 MPPS), 913201164 bytes (7.616486 Gbps)                 
    Loss: 0 ingress drop + 11319 packets lost (0.694217%)            
```

dpino commented:
Actually I think there's something wrong because at a 9.6 Gbps load with 0 packet drops the throughput should be higher than 1.6 MPPS :/

dpino commented:
Commands:

```
$ sudo ./snabb lwaftr run --cpu 11 --conf lwaftr.conf --name lwaftr --v4 82:00.0 --v6 82:00.1
```

```
$ sudo ./snabb loadtest transient --cpu 2 -D 1 -b 10e9 -s 0.2e9 from-inet-test.pcap "NIC 0" "NIC 1" 02:00.0 from-b4-test.pcap "NIC 1" "NIC 0" 02:00.1
```

dpino commented:
I run the current lwAFTR branch (same commands), the results are not better :/

```
Applying 8.800000 Gbps of load.             
  NIC 0:                                    
    TX 1578246 packets (1.578246 MPPS), 890130744 bytes (7.424069 Gbps)                 
    RX 1446140 packets (1.446140 MPPS), 873468560 bytes (7.265407 Gbps)                 
    Loss: 374 ingress drop + 131732 packets lost (8.370431%)                            
  NIC 1:                                    
    TX 1578246 packets (1.578246 MPPS), 953260584 bytes (7.929108 Gbps)                 
    RX 1446103 packets (1.446103 MPPS), 815602092 bytes (6.802469 Gbps)                 
    Loss: 349 ingress drop + 131794 packets lost (8.372776%)                            
Applying 9.000000 Gbps of load.             
  NIC 0:                                    
    TX 1556418 packets (1.556418 MPPS), 877819752 bytes (7.321390 Gbps)                 
    RX 1477216 packets (1.477216 MPPS), 892238464 bytes (7.421533 Gbps)                 
    Loss: 172 ingress drop + 79030 packets lost (5.088736%)                             
  NIC 1:                                    
    TX 1556418 packets (1.556418 MPPS), 940076472 bytes (7.819444 Gbps)                 
    RX 1477276 packets (1.477276 MPPS), 833183664 bytes (6.949106 Gbps)                 
    Loss: 138 ingress drop + 79004 packets lost (5.084881%)                             
Applying 9.200000 Gbps of load.             
  NIC 0:                                    
    TX 1705134 packets (1.705134 MPPS), 961695576 bytes (8.020950 Gbps)                 
    RX 1406171 packets (1.406171 MPPS), 849327284 bytes (7.064603 Gbps)                 
    Loss: 0 ingress drop + 298963 packets lost (17.533109%)                             
  NIC 1:                                    
    TX 1705134 packets (1.705134 MPPS), 1029900936 bytes (8.566593 Gbps)                
    RX 1406179 packets (1.406179 MPPS), 793084956 bytes (6.614666 Gbps)                 
    Loss: 0 ingress drop + 298955 packets lost (17.532640%)
```

dpino commented:
It seems the bad performance issues were due to the host where the benchmark was run. Using another host I got good results for this branch:

```
$ sudo ./snabb lwaftr run --name lwaftr --cpu 5 --conf lwaftr-datasets/lwaftr-1e6-sw/test.conf --v4 03:00.0 --v6 03:00.1
```

```
$ sudo ./snabb loadtest transient --cpu=11 -D 2 -b 10e9 -s 0.2e9 lwaftr-datasets/lwaftr-1e6-sw/from-inet-test.pcap v4 v6 83:00.0 lwaftr-datasets/lwaftr-1e6-sw/from-b4-test.pcap v6 v4 83:00.1
```

```
Applying 9.000000 Gbps of load.             
  v4:                                       
    TX 3826526 packets (1.913263 MPPS), 2158160664 bytes (8.999989 Gbps)                
    RX 3826526 packets (1.913263 MPPS), 2311221704 bytes (9.612233 Gbps)                
    Loss: 0 ingress drop + 0 packets lost (0.000000%)                                   
  v6:                                       
    TX 3582798 packets (1.791399 MPPS), 2164009992 bytes (8.999989 Gbps)                
    RX 3582798 packets (1.791399 MPPS), 2020698072 bytes (8.426741 Gbps)                
    Loss: 0 ingress drop + 0 packets lost (0.000000%)                                   
Applying 9.200000 Gbps of load.             
  v4:                                       
    TX 3911525 packets (1.955763 MPPS), 2206100100 bytes (9.199907 Gbps)                
    RX 3908705 packets (1.954352 MPPS), 2360857820 bytes (9.818667 Gbps)                
    Loss: 0 ingress drop + 2820 packets lost (0.072095%)                                
  v6:                                       
    TX 3662383 packets (1.831192 MPPS), 2212079332 bytes (9.199906 Gbps)                
    RX 3662383 packets (1.831192 MPPS), 2065584012 bytes (8.613925 Gbps)                
    Loss: 0 ingress drop + 0 packets lost (0.000000%)                         
```

dpino committed:
Temporarily disable ipsec selftest

dpino commented:
It seems to be an issue with snabbvmx's nexthop selftest.

Tests run locally (all tests that depend on NICs are skipped):

```
TEST      core.app
TEST      core.histogram
TEST      core.worker
TEST      core.link
TEST      core.timer
TEST      core.memory
TEST      core.lib
TEST      core.packet
TEST      core.shm
TEST      core.main
TEST      core.counter
TEST      apps.bridge.mac_table
TEST      apps.intel.intel_app
SKIPPED   testlog/apps.intel.intel_app
TEST      apps.ipv6.fragment
TEST      apps.ipv6.reassemble
TEST      apps.ipv6.nd_light
TEST      apps.vhost.vhost_user
SKIPPED   testlog/apps.vhost.vhost_user
TEST      apps.packet_filter.pcap_filter
TEST      apps.keyed_ipv6_tunnel.tunnel
TEST      apps.ipfix.ipfix
TEST      apps.ipfix.template
TEST      apps.test.synth
TEST      apps.test.match
TEST      apps.virtio_net.virtio_net
SKIPPED   testlog/apps.virtio_net.virtio_net
TEST      apps.rate_limiter.rate_limiter
TEST      apps.lwaftr.lwdebug
TEST      apps.lwaftr.nh_fwd
TEST      apps.lwaftr.V4V6
TEST      apps.lwaftr.ctable_wrapper
TEST      apps.lwaftr.ndp
TEST      apps.lwaftr.binding_table
TEST      apps.lwaftr.rangemap
TEST      apps.tap.tap
SKIPPED   testlog/apps.tap.tap
TEST      apps.example.asm
TEST      apps.vlan.vlan
TEST      apps.socket.unix
TEST      apps.socket.raw
TEST      apps.wall.scanner
TEST      apps.wall.l7fw
TEST      apps.ipv4.fragment
TEST      apps.ipv4.reassemble
TEST      apps.ipv4.arp
TEST      program.l2vpn.control_channel
TEST      program.l2vpn.pseudowire
TEST      program.lwaftr.quickcheck.utils
TEST      program.lwaftr.tests.propbased.genyang
TEST      program.snabbnfv.nfvconfig
SKIPPED   testlog/program.snabbnfv.nfvconfig
TEST      program.snabbnfv.neutron2snabb.neutron2snabb_schema
TEST      program.snabbnfv.neutron2snabb.neutron2snabb
TEST      test_schema
ERROR     testlog/test_schema
TEST      lib.binary_search
TEST      lib.hash.murmur
TEST      lib.hash.siphash
TEST      lib.checksum
TEST      lib.traceprof.traceprof
TEST      lib.lpm.lpm
TEST      lib.lpm.lpm4_248
TEST      lib.lpm.lpm4
TEST      lib.lpm.lpm4_poptrie
TEST      lib.lpm.lpm4_dxr
TEST      lib.lpm.ip6
TEST      lib.lpm.lpm4_trie
TEST      lib.lpm.ip4
TEST      lib.ctable
TEST      lib.pmu_x86
TEST      lib.numa
TEST      lib.ipsec.esp
TEST      lib.ipsec.aes_128_gcm
TEST      lib.scheduling
TEST      lib.multi_copy
TEST      lib.ipc.shmem.shmem
TEST      lib.cltable
TEST      lib.protocol.datagram
TEST      lib.protocol.ipv6
TEST      lib.protocol.tcp
TEST      lib.protocol.ipv4
TEST      lib.yang.value
TEST      lib.yang.path
TEST      lib.yang.util
TEST      lib.yang.parser
TEST      lib.yang.alarms
TEST      lib.yang.rpc
TEST      lib.yang.state
TEST      lib.yang.path_data
TEST      lib.yang.schema
TEST      lib.yang.data
TEST      lib.yang.yang
TEST      lib.yang.binary
TEST      lib.pmu
SKIPPED   testlog/lib.pmu
TEST      lib.hardware.pci
TEST      lib.ptree.json
TEST      lib.ptree.worker
TEST      lib.ptree.channel
TEST      lib.ptree.alarm_codec
TEST      lib.ptree.trace
TEST      lib.ptree.action_codec
TEST      lib.ptree.ptree
TEST      selftest.sh
TEST      apps/solarflare/selftest.sh
SKIPPED   testlog/apps.solarflare.selftest.sh
TEST      apps/ipsec/selftest.sh
SKIPPED   testlog/apps.ipsec.selftest.sh
TEST      apps/intel_mp/selftest.sh
SKIPPED   testlog/apps.intel_mp.selftest.sh
TEST      apps/tap/selftest.sh
TEST      program/snabbvmx/tests/end-to-end/selftest.sh
TEST      program/snabbvmx/tests/nexthop/selftest.sh
SKIPPED   testlog/program.snabbvmx.tests.nexthop.selftest.sh
TEST      program/packetblaster/selftest.sh
SKIPPED   testlog/program.packetblaster.selftest.sh
TEST      program/lwaftr/tests/selftest.sh
TEST      program/lwaftr/tests/end-to-end/selftest.sh
TEST      program/lwaftr/tests/soaktest/selftest.sh
TEST      program/lwaftr/tests/config-migrations/selftest.sh
TEST      program/lwaftr/tests/propbased/selftest.sh
SKIPPED   testlog/program.lwaftr.tests.propbased.selftest.sh
TEST      program/snabbnfv/selftest.sh
TEST      program/snabbnfv/neutron2snabb/selftest.sh
TEST      program/wall/tests/selftest.sh
SKIPPED   testlog/program.wall.tests.selftest.sh
TEST      lib/watchdog/selftest.sh
```

wingo commented on src/program/snabbvmx/tests/nexthop/selftest.sh:
```diff
@@ -54,7 +57,8 @@ fi
 COUNT=0
 while true; do
-    output=$(./snabb snabbvmx query | grep "next_hop_mac")
+    $(./snabb snabbvmx query > $TMPFILE)
```
I don't really understand how this could make anything better.  If we keep the change then the `$()` on this line are not needed.

dpino replied:
I will remove this patch as it was properly fixed in #1031

wingo committed:
Make supervisor setup more reliable

There was a race condition when setting up the supervisor such that in
some cases it was possible to miss a signal when the parent process
died.  We could reproduce this with by running a "snabb lwaftr bench",
but only on our test machine with two NUMA nodes and only when setting
--cpu on the lwaftr.  In that case the problem would appear when
running "snabb lwaftr monitor" on the lwaftr; the monitor process
would hang reading from the signalfd.  Because the monitor process
still had stdout open, then when piping its output to "grep", the grep
process would hang because the write side of its stdin pipe would
still be open as well.

This patch fixes this error by making cleanup reliable.  It does so by
taking a POSIX lock on an unnamed file in the parent, then taking
another lock from the supervisor child process.  In this way we avoid
some of the more arcane parts of Linux.

wingo commented:
Regarding ipsec, @eugeneia comments on Slack that "[these tests] are disabled in davos as well. It depends on extended sequence numbers that isn't in older kernels. Should have probably made the env var a positive ON switch instead."

So, disabling ipsec tests is the right thing.  I think we can proceed to a release.

wingo commented on .version:
```diff
@@ -1 +1 @@
-2017.11.01
```
One final comment here -- since this release is based on Snabb v2018.01 it should be Snabb lwAFTR v2018.01.XXXX, not .03.  Can we change to v2018.01.2.1 or something?  I know that having the four version components is not great but given that upstream had to make a .1 and a .2 release, it seems reasonable.

dpino committed:
Snabb lwAFTR v2018.01.2.1
